### ISSUE TITLE
FR: better error message on dimension mismatch 

### ISSUE THREAD

**Issue by mike-lawrence** [stan-dev/stanc3#960]
With cmdstan 2.27.0, the following stan code:
```stan
data{
	int n;
	int m;
	matrix[n,m] x ;
}
transformed data{
	matrix[n,m] z = transpose(x) ; //bug; should be `matrix[m,n] ...`
}
```
passes syntax checks (inc. pendantic) but yields the unhelpful error at runtime:
```
stan/lib/stan_math/lib/eigen_3.3.9/Eigen/src/Core/DenseBase.h:256: void Eigen::DenseBase<Derived>::resize(Eigen::Index, Eigen::Index) [with Derived = Eigen::Map<Eigen::Matrix<double, -1, -1> >; Eigen::Index = long int]: Assertion `rows == this->rows() && cols == this->cols() && "DenseBase::resize() does not actually allow to resize."' failed.
```
Would there be any avenue for a more informative error message calling out the line number in the original stan code where the dimension mismatch arises?

**Comment by WardBrian** [comment]
I think this would need to be handled on the c++ side of things. The simplest fix is probably redefining [`eigen_assert`](https://eigen.tuxfamily.org/dox/TopicAssertions.html) so that it throws an exception which is then caught and 'relocated' by the existing mechanism we have for error messages, but it may also be worth inspecting why this assertion gets hit and if we can avoid it.

**Comment by rok-cesnovar** [comment]
Yeah, this should be handled in assign in Stan C++ (moving this issue there). 

This assign is called: https://github.com/stan-dev/stan/blob/ee883ebdf500a0832df93878de73698226b2127c/src/stan/model/indexing/assign.hpp#L57

And this one doesnt have a size check that the other assigns below have. I guess it should have one?

cc: @SteveBronder @t4c1 if any one of you have thoughts on this.
